File: Project_Absenteeism.py
# -*- coding: utf-8 -*-
"""
Created on Mon Aug 13 18:22:34 2018

@author: Harshal
"""

# Import Libraries
import os
import numpy as np
import pandas as pd
from fancyimpute import KNN

# Set active Work Directory
os.chdir("C:/Users/Harshal/Desktop/Edwisor/Project/Project_Datafiles")

# Load Data
xls = pd.ExcelFile("Absenteeism_at_work_Project.xls")
EmployeeData = xls.parse()

# Get Cnames
def get_cname(data):
    all_cnames = []
    num_cnames = []
    cat_cnames = []
    for i in data.columns:
        all_cnames.append(str(i))
        if(data[i].dtype == "object"):
            cat_cnames.append(str(i))
        else:
            num_cnames.append(str(i))
    cnames = [all_cnames, num_cnames, cat_cnames]    
    return(cnames)

# Get cnames - cnames[0] - all cnames; cnames[1] - numeric cnames; cnames[2] - categorical cnames 
cnames = get_cname(EmployeeData)

#rows, cols = EmployeeData.shape
rows = EmployeeData.shape[0] #gives number of row count
cols = EmployeeData.shape[1] #gives number of col count

# Exploratory Data Analysis 
for i in cnames[0]:
    print(str(i) + "    " + str(type(EmployeeData[i][1])))
    
# Change Data as per problem requirement  
    for i in range(0,rows):
        if EmployeeData["Absenteeism time in hours"][i] != 0:
            if EmployeeData["Reason for absence"][i] == 0:
                EmployeeData["Reason for absence"][i] = np.nan
            if EmployeeData["Month of absence"][i] == 0:
                EmployeeData["Month of absence"][i] = np.nan
        
def preprocessing(EmployeeData):
    # Change into require Data Types
    EmployeeData["ID"] = EmployeeData["ID"].astype(str)
    EmployeeData["Reason for absence"] = EmployeeData["Reason for absence"].astype(str)
    EmployeeData["Month of absence"] = EmployeeData["Month of absence"].astype(str)
    EmployeeData["Day of the week"] = EmployeeData["Day of the week"].astype(str)
    EmployeeData["Seasons"] = EmployeeData["Seasons"].astype(str)
    EmployeeData["Disciplinary failure"] = EmployeeData["Disciplinary failure"].astype(str)
    EmployeeData["Education"] = EmployeeData["Education"].astype(str)
    EmployeeData["Son"] = EmployeeData["Son"].astype(str)
    EmployeeData["Social drinker"] = EmployeeData["Social drinker"].astype(str)
    EmployeeData["Social smoker"] = EmployeeData["Social smoker"].astype(str)
    EmployeeData["Pet"] = EmployeeData["Pet"].astype(str)
    
    # Change NaN string values back to NaN 
    EmployeeData["ID"] = EmployeeData["ID"].replace("nan",np.nan)
    EmployeeData["Reason for absence"] = EmployeeData["Reason for absence"].replace("nan",np.nan)
    EmployeeData["Month of absence"] = EmployeeData["Month of absence"].replace("nan",np.nan)
    EmployeeData["Day of the week"] = EmployeeData["Day of the week"].replace("nan",np.nan)
    EmployeeData["Seasons"] = EmployeeData["Seasons"].replace("nan",np.nan)
    EmployeeData["Disciplinary failure"] = EmployeeData["Disciplinary failure"].replace("nan",np.nan)
    EmployeeData["Education"] = EmployeeData["Education"].replace("nan",np.nan)
    EmployeeData["Son"] = EmployeeData["Son"].replace("nan",np.nan)
    EmployeeData["Social drinker"] = EmployeeData["Social drinker"].replace("nan",np.nan)
    EmployeeData["Social smoker"] = EmployeeData["Social smoker"].replace("nan",np.nan)
    EmployeeData["Pet"] = EmployeeData["Pet"].replace("nan",np.nan)
    
    # Covert factor varaible values to labels 
    for i in range(0, len(EmployeeData.columns)):
        if(EmployeeData.iloc[:,i].dtypes == 'object'):
            EmployeeData.iloc[:,i] = pd.Categorical(EmployeeData.iloc[:,i])
            EmployeeData.iloc[:,i] = EmployeeData.iloc[:,i].cat.codes 
            EmployeeData.iloc[:,i] = EmployeeData.iloc[:,i].astype('object')
    
    # Convert -1 values back to NaN
    for i in cnames[0]:
        for j in range(0,rows):
            if EmployeeData.loc[j,i] == -1:
                EmployeeData.loc[j,i] = np.nan
    
    return EmployeeData

EmployeeData = preprocessing(EmployeeData)
cnames = get_cname(EmployeeData)

# Missing Value analysis
def missingValueCheck(data):
    print(data.isna().sum())
missingValueCheck(EmployeeData)

# Impute values related to ID
Depenedent_ID =  ["ID","Transportation expense","Service time","Age","Height",
                   "Distance from Residence to Work","Education","Son","Weight",
                   "Social smoker","Social drinker","Pet","Body mass index"]
Depenedent_ID_data = EmployeeData[Depenedent_ID].copy()
Depenedent_ID_data = Depenedent_ID_data.groupby("ID").max()
Depenedent_ID.remove('ID')

for i in Depenedent_ID:
    for j in range(0,rows):
        RI = EmployeeData["ID"][j]
        if np.isnan(EmployeeData.loc[j,i]):
            EmployeeData[i][j] = Depenedent_ID_data.loc[RI,i]
            
# Impute values for other variables with KNN            
EmployeeData = pd.DataFrame(KNN(k = 7).complete(EmployeeData), columns = EmployeeData.columns)
EmployeeData = preprocessing(EmployeeData)
missingValueCheck(EmployeeData)

# Outlier Analysis
# Get cnames of numeric varaibles not dependent on ID
numeric_cnames = ['Work load Average/day ','Hit target']

# Impute Outliers with NA
for i in numeric_cnames:
    q75, q25 = np.nanpercentile(EmployeeData.loc[:,i],[75, 25])
    iqr = q75 - q25
    min = q25 - (iqr*1.5)
    max = q75 + (iqr*1.5)
    EmployeeData.loc[EmployeeData[i] < min, i] = np.nan
    EmployeeData.loc[EmployeeData[i] > max, i] = np.nan
#Impute with KNN
EmployeeData = pd.DataFrame(KNN(k = 7).complete(EmployeeData), columns = EmployeeData.columns)
EmployeeData = preprocessing(EmployeeData)
missingValueCheck(EmployeeData)

# Feature Selection
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_regression
X = EmployeeData.iloc[:,:-1].values
y = EmployeeData.iloc[:,20].values
# Create an SelectKBest object to select features with two best ANOVA F-Values
selector = SelectKBest(f_regression, k=7)
# Apply the SelectKBest object to the features and target
X = selector.fit_transform(X, y)
selector.scores_

# Feature Scaling
#Nomalisation
from sklearn.preprocessing import normalize
X = normalize(X, norm='l2')
#for i in range(0,(X_kbest.shape[1]-1)):
#    X_kbest[i] = (X_kbest[i] - np.min(X_kbest[i])) / (np.max(X_kbest[i]) - np.min(X_kbest[i]))

# Error Matrix
from sklearn.metrics import mean_squared_error
from math import sqrt
def RMSE(y, pred):
    print(sqrt(mean_squared_error(y, pred)))
    
def MSE(y, pred):
    print(mean_squared_error(y, pred))

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Multiple Linear Regression
from sklearn.linear_model import LinearRegression
lm_regressor = LinearRegression()
lm_regressor.fit(X_train, y_train)
lm_predict = lm_regressor.predict(X_test)
RMSE(y_test, lm_predict)
MSE(y_test, lm_predict)
     
# Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor
DT_regressor = DecisionTreeRegressor()
DT_regressor.fit(X_train, y_train)
DT_predict = DT_regressor.predict(X_test)
RMSE(y_test, DT_predict)
MSE(y_test, DT_predict)

# Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor
RF_regressor = RandomForestRegressor()
RF_regressor.fit(X_train, y_train)
RF_predict = RF_regressor.predict(X_test)
RMSE(y_test, RF_predict)
MSE(y_test, RF_predict)

# Support Vector Regressor
from sklearn.svm import SVR
SVR_regressor = SVR(kernel='rbf')
SVR_regressor.fit(X_train, y_train)
SVR_predict = SVR_regressor.predict(X_test)
RMSE(y_test, SVR_predict)
MSE(y_test, SVR_predict)

############################## Problems ##################################
# Suggesting the changes
RF_regressor_p1 = RandomForestRegressor().fit(X, y)
RF_regressor_p1.feature_importances_

# Calculating Losses
p2_data = X
#Predict for new test cases
p2_predict = RF_regressor.predict(p2_data)
p2_predict = pd.DataFrame(p2_predict)
p2_predict.columns = ['predictions']
# Add predicted values to the DataSet
p2_frames = [EmployeeData, p2_predict]
p2_dataSet = pd.concat(p2_frames, axis=1)

# Calculate the total Loss 
Loss = 0
for i in range(0,p2_dataSet.shape[0]):
    if (p2_dataSet['Hit target'][i] != 100):
        if (p2_dataSet['Age'][i] >= 25 and  p2_dataSet['Age'][i] <= 32):
            Loss += (p2_dataSet['Disciplinary failure'][i] + 1) * 2000 + (p2_dataSet['Education'][i] + 1 + 1) * 500 + p2_dataSet['predictions'][i] * 1000
        elif (p2_dataSet['Age'][i] >= 33 and  p2_dataSet['Age'][i] <= 40):
            Loss += (p2_dataSet['Disciplinary failure'][i] + 1) * 2000 + (p2_dataSet['Education'][i] + 1 + 2) * 500 + p2_dataSet['predictions'][i] * 1000
        elif (p2_dataSet['Age'][i] >= 41 and  p2_dataSet['Age'][i] <= 49):
            Loss += (p2_dataSet['Disciplinary failure'][i] + 1) * 2000 + (p2_dataSet['Education'][i] + 1 + 3) * 500 + p2_dataSet['predictions'][i] * 1000
        elif (p2_dataSet['Age'][i] >= 50 and  p2_dataSet['Age'][i] <= 60):
            Loss += (p2_dataSet['Disciplinary failure'][i] + 1) * 2000 + (p2_dataSet['Education'][i] + 1 + 4) * 500 + p2_dataSet['predictions'][i] * 1000

# To calculate loss per month
Loss = Loss/12
print(Loss)

File: Project_Absenteeism.R
# Clean current environment
rm(list=ls())
# Set work directory
setwd("C:/Users/Harshal/Desktop/Edwisor/Project/Project_Datafiles")

# Load require Packages
p <- c("xlsx","DMwR","corrgram","caret","usdm","rpart","DataCombine","randomForest",
       "e1071")
lapply(p, require, character.only=TRUE)
rm(p)

EmployeeData <- read.xlsx("Absenteeism_at_work_Project.xls", sheetIndex = 1)

###################### Exploratory Data Analysis ##################
# Check variable types
cnames <- colnames(EmployeeData)

# Check number of unique variables
for (i in cnames){
  print(i)
  print(aggregate(data.frame(count = EmployeeData[,i]), 
                  list(value = EmployeeData[,i]), length))
}

# Data Preprocessing 
preprocessing <- function(EmployeeData){
  EmployeeData$ID <- as.factor(EmployeeData$ID)
  for (i in (1:nrow(EmployeeData))){
    if (EmployeeData$Absenteeism.time.in.hours[i] != 0 || is.na(EmployeeData$Absenteeism.time.in.hours[i])){
      if(EmployeeData$Reason.for.absence[i] == 0 || is.na(EmployeeData$Reason.for.absence[i])){
        EmployeeData$Reason.for.absence[i] = NA
      }
      if(EmployeeData$Month.of.absence[i] == 0 || is.na(EmployeeData$Month.of.absence[i])){
        EmployeeData$Month.of.absence[i] = NA
      }
    }
  }
  EmployeeData$Reason.for.absence <- as.factor(EmployeeData$Reason.for.absence)
  EmployeeData$Month.of.absence <- as.factor(EmployeeData$Month.of.absence)
  EmployeeData$Day.of.the.week <- as.factor(EmployeeData$Day.of.the.week)
  EmployeeData$Seasons <- as.factor(EmployeeData$Seasons)
  EmployeeData$Disciplinary.failure <- as.factor(EmployeeData$Disciplinary.failure)
  EmployeeData$Education <- as.factor(EmployeeData$Education)
  EmployeeData$Son <- as.factor(EmployeeData$Son)
  EmployeeData$Social.drinker <- as.factor(EmployeeData$Social.drinker)
  EmployeeData$Social.smoker <- as.factor(EmployeeData$Social.smoker)
  EmployeeData$Pet <- as.factor(EmployeeData$Pet)
  return(EmployeeData)
}
EmployeeData <- preprocessing(EmployeeData)

#selecting only factor
get_cat <- function(data) {
  return(colnames(data[,sapply(data, is.factor)]))
}
cat_cnames <- get_cat(EmployeeData)

#selecting only numeric
get_num <- function(data) {
  return(colnames(data[,sapply(data, is.numeric)]))
}
num_cnames <- get_num(EmployeeData)

# Covert factor varaible values to labels 
for (i in cat_cnames){
  EmployeeData[,i] <- factor(EmployeeData[,i], 
                             labels = (1:length(levels(factor(EmployeeData[,i])))))
}

############################### Data PreProcessing #########################
# Missing Value Analysis
# Get Missing Values for all Variables
missingValueCheck <- function(data){
  for (i in colnames(data)){
    print(i)
    print(sum(is.na(data[i])))
  }
  print("Total")
  print(sum(is.na(EmployeeData)))
}
missingValueCheck(EmployeeData)

# Impute values related to ID
Depenedent_ID <- c("ID","Transportation.expense","Service.time","Age","Height",
                   "Distance.from.Residence.to.Work","Education","Son","Weight",
                   "Social.smoker","Social.drinker","Pet","Body.mass.index")
Depenedent_ID_data <- EmployeeData[,Depenedent_ID]
Depenedent_ID_data <- aggregate(. ~ ID, data = Depenedent_ID_data, 
                     FUN = function(e) c(x = mean(e)))
for (i in Depenedent_ID) {
  for (j in (1:nrow(EmployeeData))){
    ID <- EmployeeData[j,"ID"]
    if(is.na(EmployeeData[j,i])){
      EmployeeData[j,i] <- Depenedent_ID_data[ID,i]
    }
  }
}
# Impute values for other variables
EmployeeData = knnImputation(EmployeeData, k = 7)
missingValueCheck(EmployeeData)

# Outlier Analysis
# Histogram
for(i in num_cnames){
  hist(EmployeeData[,i], xlab=i, main=" ", col=(c("lightblue","darkgreen")))
}

# BoxPlot
num_cnames <- num_cnames[ num_cnames != "Absenteeism.time.in.hours"]
for(i in num_cnames){
  boxplot(EmployeeData[,i]~EmployeeData$Absenteeism.time.in.hours,
          data=EmployeeData, main=" ", ylab=i, xlab="Absenteeism.time.in.hours",  
          col=(c("lightblue","darkgreen")), outcol="red")
}

# Removing ID related varaibles from num_cnames
for (i in Depenedent_ID){
  num_cnames <- num_cnames[ num_cnames != i]
}
# Replace all outliers with NA and impute
for(i in num_cnames){
  val = EmployeeData[,i][EmployeeData[,i] %in% boxplot.stats(EmployeeData[,i])$out]
  EmployeeData[,i][EmployeeData[,i] %in% val] = NA
}

# Impute Missing Values
missingValueCheck(EmployeeData)
EmployeeData <- knnImputation(EmployeeData, k = 7)

# Copy Employee Data into DataSet for further analysis
DataSet <- EmployeeData

######################## Fature Selection #################
# Correlation Plot
corrgram(EmployeeData, upper.panel=panel.pie, text.panel=panel.txt, 
         main = "Correlation Plot")

# ANOVA
cnames <- colnames(EmployeeData)
for (i in cnames){
  print(i)
  print(summary(aov(EmployeeData$Absenteeism.time.in.hours~EmployeeData[,i], 
                    EmployeeData)))
}

# Dimensionality Reduction
EmployeeData <- subset(EmployeeData, select = -c(Weight,Education,Service.time,
                                                 Social.smoker,Body.mass.index,
                                                 Work.load.Average.day.,Seasons,
                                                 Transportation.expense,Pet,
                                                 Disciplinary.failure,Hit.target,
                                                 Month.of.absence,Social.drinker))

######################### Fature Normalization ######################
cat_cnames <- get_cat(EmployeeData)
num_cnames <- get_num(EmployeeData)

# Fature Scaling
for (i in num_cnames){
  EmployeeData[,i] <- (EmployeeData[,i] - min(EmployeeData[,i])) / 
                        (max(EmployeeData[,i]) - min(EmployeeData[,i]))
}

#################### Model Development ########################
# Data Divide
X_index <- sample(1:nrow(EmployeeData), 0.8 * nrow(EmployeeData))
X_train <- EmployeeData[X_index,-8]
X_test <- EmployeeData[-X_index,-8]
y_train <- EmployeeData[X_index,8]
y_test <- EmployeeData[-X_index,8]

train <- EmployeeData[X_index,]
test <- EmployeeData[-X_index,]

#calculate RMSE
RMSE <- function(y, yhat){
  sqrt(mean((y - yhat)^2))
}

#calculate MSE
MSE <- function(y, yhat){
  (mean((y - yhat)^2))
}

###################### Multiple Linear Regression ########################
num_data <- train[sapply(train, is.numeric)]
vifcor(num_data, th=0.9)

lm_regressor <- lm(Absenteeism.time.in.hours~.,data = train)
summary(lm_regressor)
#Predict for new test cases
for (i in cat_cnames){
  lm_regressor$xlevels[[i]] <- union(lm_regressor$xlevels[[i]], 
                                        levels(X_test[[i]]))
}
lm_predict <- predict(lm_regressor,newdata=X_test)
RMSE(lm_predict,y_test)
MSE(lm_predict,y_test)

###############################Decision Trees for regression ###################
DT_regressor <- rpart(Absenteeism.time.in.hours~.,data = train, method="anova")
#Predict for new test cases
DT_predict <- predict(DT_regressor, X_test)
RMSE(DT_predict,y_test)
MSE(DT_predict,y_test)

##################### Random Forest ########################
RF_regressor <- randomForest(x = X_train, y = y_train, ntree = 100)
#Predict for new test cases
RF_predict <- predict(RF_regressor, X_test)
RMSE(RF_predict,y_test)
MSE(RF_predict,y_test)

#################### Support Vector Regressor #####################
SVR_regressor <- svm(formula = Absenteeism.time.in.hours ~ ., 
                data = train, type = 'eps-regression')
#Predict for new test cases
SVR_predict <- predict(SVR_regressor, X_test)
RMSE(SVR_predict,y_test)
MSE(SVR_predict,y_test)

############################## Problems ##################################
# Suggesting the changes
lm_regressor_p1 <- lm(Absenteeism.time.in.hours~.,data = EmployeeData)
summary(lm_regressor_p1)

# Calculating Losses
p2_data = EmployeeData[,-8]
#Predict for new test cases
p2_predict <- predict(RF_regressor, p2_data)

# Convert predict values back to original scale
p2_predict <- (p2_predict * 120)

# Add predicted values to the DataSet
p2_dataSet <- merge(DataSet,p2_predict,by="row.names",all.x=TRUE)

# Calculate the total Loss 
Loss <- 0
for (i in 1:nrow(p2_dataSet)){
  if (p2_dataSet$Hit.target[i] != 100)
    if (p2_dataSet$Age[i] >= 25 &&  p2_dataSet$Age[i] <= 32){
      Loss = Loss + as.numeric(p2_dataSet$Disciplinary.failure[i]) * 2000 +
        (as.numeric(p2_dataSet$Education[i]) + 1) * 500 + p2_dataSet$y[i] * 1000
    }else if(p2_dataSet$Age[i] >= 33 &&  p2_dataSet$Age[i] <= 40){
      Loss = Loss + as.numeric(p2_dataSet$Disciplinary.failure[i]) * 2000 +
        (as.numeric(p2_dataSet$Education[i]) + 2) * 500 + p2_dataSet$y[i] * 1000
    }else if(p2_dataSet$Age[i] >= 41 &&  p2_dataSet$Age[i] <= 49){
      Loss = Loss + as.numeric(p2_dataSet$Disciplinary.failure[i]) * 2000 +
        (as.numeric(p2_dataSet$Education[i]) + 3) * 500 + p2_dataSet$y[i] * 1000
    }else if(p2_dataSet$Age[i] >= 50 &&  p2_dataSet$Age[i] <= 60){
      Loss = Loss + as.numeric(p2_dataSet$Disciplinary.failure[i]) * 2000 +
        (as.numeric(p2_dataSet$Education[i]) + 4) * 500 + p2_dataSet$y[i] * 1000
    } 
}
# To calculate loss per month
Loss <- Loss/12
Loss
